[Console] Expose the original input arguments and options and to unparse options

theofidry opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 8.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | None
| License       | MIT

Adds `RawInputInterface` (extending `InputInterface`) with three methods to support forwarding parsed input to child processes:

- `getRawArguments()`: returns arguments without default values merged in
- `getRawOptions()`: returns options without default values merged in
- `unparse(?array $optionNames = null): array`: converts parsed options back to their CLI representation (`['--opt=value', '--flag']`), suitable for passing to `Process`

`Input` (the base class for `ArgvInput`, `ArrayInput`, `StringInput`) implements the new interface. The interface is also wired into the argument resolver, invokable commands, DI, and `SymfonyRuntime` so that commands can type-hint `RawInputInterface` directly.

### Use case

Forwarding the current command's input to a child process, e.g. in [console-parallelization](https://github.com/webmozarts/console-parallelization):

```php
$options = $input->getRawOptions();
unset($options['main-process-only-option']);
$options['child'] = true;

$process = new Process([
    PHP_BINARY, 'bin/console', 'my:command',
    ...$input->getRawArguments(),
    ...$input->unparse(array_keys($options)),
]);
```

### Design notes

- **No BC break**: `InputInterface` is untouched. `RawInputInterface extends InputInterface` so type-hinting it gives access to the full input API.
- **`unparse()` returns `list<string>`** (not a concatenated string) so each element maps to a single process argument — no double-escaping.
- **Default values are excluded** from `getRawArguments()`/`getRawOptions()` so only explicitly passed input is forwarded.

theofidry committed:
[Console] Expose the original input arguments and options and to unparse options

theofidry commented:
@chalasr updated

theofidry commented on src/Symfony/Component/Console/Input/Input.php:
```diff
@@ -13,6 +13,7 @@
 
 use Symfony\Component\Console\Exception\InvalidArgumentException;
 use Symfony\Component\Console\Exception\RuntimeException;
+use function array_keys;
```
unsure about those imports, but I thought the CS fixer was taking care of those?

welcoMattic replied:
It's better to prefix your `array_keys` call by `\` to avoid those imports

theofidry commented on src/Symfony/Component/Console/Input/Input.php:
```diff
@@ -85,6 +86,35 @@ public function getArguments(): array
     }
 
+    /**
+     * Returns all the given arguments NOT merged with the default values.
```
I do not know if the description should be kept. I don't think I can add a strict equivalent when using `@method` to `InputInterface` hence might be worth to keep it there for having the complete doc when adding the method to the interface instead

theofidry commented on src/Symfony/Component/Console/Input/Input.php:
```diff
@@ -85,6 +86,35 @@ public function getArguments(): array
+     *
+     * @return array<string|bool|int|float|null|array<string|bool|int|float|null>>
+     */
+    public function getRawArguments(bool $strip = false): array
```
I added `bool $strip = false` to keep the parity with `getRawTokens()`, but I don't really need it so I don't have an opinion as to whether it should be kept

OskarStark replied:
Just my 2 cents, but if we don't use it, it should be removed

cc @chalasr

chalasr commented:
Should the `getRawX()` methods introduced here replace the `ArgvInput::getRawTokens()`? 
/cc @lyrixx - does they work for castor's usecase?

As discussed, I'm not entirely convinced we need `unparse()`:
- `parse()` is not defined by `InputInterface`
- could the mentioned use case be solved through manipulating the retval of `__toString()`?
- overall, use cases for parse()/unparse() are marginal, so is it worth it?

theofidry commented:
>Should the `getRawX()` methods introduced here replace the `ArgvInput::getRawTokens()`?

They could, but to this effect if what you're looking for is what `getRawTokens()` provides, then it's simpler than using `getRawX()`.

>parse() is not defined by InputInterface

`parse()` is not, but it's caller, and why this unparse is needed, is `InputInterface::bind()`, it's just that `::unbind()` be closer to a `reset()` IMO.

>could the mentioned use case be solved through manipulating the retval of __toString()?

As highlighted in the description, they cannot, at least not with the same level of sketchiness as the workaround currently employed which as listed in the "user-land alternative" section. Rather than manipulating a string, I would rather use raw tokens, but that is only available for `ArgvInput` and you still need to work out manipulating the tokens based on the input option names and types (e.g. negatable).

>overall, use cases for parse()/unparse() are marginal, so is it worth it?

Can't decide on that, but to fix it "correctly" it is certainly the right place. Personally I find weird that the framework does not provide the primitive to reverse the transformation it does.

OskarStark commented on src/Symfony/Component/Console/CHANGELOG.md:
```diff
@@ -1,6 +1,14 @@
+7.2
+---
+
+ * Add `InputInterface::getRawArguments()`
+ * Add `InputInterface::getRawOptions()`
+ * Add `Input::unparse()`
```
```suggestion
 * Add `InputInterface::getRawArguments()`, `InputInterface::getRawOptions()` and `Input::unparse()`
```

OskarStark commented on src/Symfony/Component/Console/Input/Input.php:
```diff
@@ -171,4 +211,55 @@ public function getStream()
+        InputOption $option,
+        string $name,
+        array|bool|float|int|string|null $value,
+    ): string {
```
one line please

OskarStark commented on src/Symfony/Component/Console/Input/Input.php:
```diff
@@ -171,4 +211,55 @@ public function getStream()
+        };
+    }
+
+    private function unparseOptionWithValue(
+        string $name,
+        bool|float|int|string|null $value,
+    ): string {
```
one line please

OskarStark commented on src/Symfony/Component/Console/Tests/Input/ArgvInputTest.php:
```diff
@@ -594,4 +594,261 @@ public static function provideGetRawTokensTrueTests(): iterable
+    /**
+     * @dataProvider unparseProvider
+     */
+    public function testUnparse(
+        ?InputDefinition $inputDefinition,
+        ArgvInput $input,
+        ?array $parsedOptions,
+        array $expected,
+    )
```
one line please

OskarStark commented on src/Symfony/Component/Console/Tests/Input/ArrayInputTest.php:
```diff
@@ -170,4 +172,278 @@ public function testToString()
+    /**
+     * @dataProvider unparseProvider
+     */
+    public function testUnparse(
```
one line please

OskarStark commented on src/Symfony/Component/Console/Tests/Input/InputTest.php:
```diff
@@ -89,10 +93,14 @@ public function testArguments()
         $this->assertEquals('default', $input->getArgument('bar'), '->getArgument() returns the default value for optional arguments');
         $this->assertEquals(['name' => 'foo', 'bar' => 'default'], $input->getArguments(), '->getArguments() returns all argument values, even optional ones');
+        $this->assertEquals(['name' => 'foo'], $input->getRawArguments(), '->getRawArguments() returns all argument values, excluding optional ones');
+        $this->assertEquals([], $input->getRawArguments(true), '->getRawArguments(true) strips the first argument value from the returned argument values');
```
Lets use `assertSame` where possible please, while you are on it, changing the other occurrences in this class is fine, thanks

theofidry replied:
Unfortunately there is side effects, so I'll address this in a separate MR

stof commented on src/Symfony/Component/Console/CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@ CHANGELOG
  * Add `OutputInterface::isSilent()`, `Output::isSilent()`, `OutputStyle::isSilent()` methods
  * Add a configurable finished indicator to the progress indicator to show that the progress is finished
  * Add ability to schedule alarm signals and a `ConsoleAlarmEvent`
+ * Add `InputInterface::getRawArguments()`, `InputInterface::getRawOptions()` and `Input::unparse()` methods
```
This should be moved to a new 7.3 section instead.

stof commented on src/Symfony/Component/Console/Input/Input.php:
```diff
@@ -171,4 +193,55 @@ public function getStream()
+        return match (true) {
+            $option->isNegatable() => \sprintf('--%s%s', $value ? '' : 'no-', $name),
+            !$option->acceptValue() => \sprintf('--%s', $name),
+            $option->isArray() => implode('', array_map(fn ($item) => $this->unparseOptionWithValue($name, $item), $value)),
```
shouldn't this use a space separator instead of an empty string ?
And shouldn't we return array options as multiple items in the unparsed list ? If you want to use this as input to symfony/process for instance, you would need each array item as a separate process argument

stof commented on src/Symfony/Component/Console/Tests/Input/ArgvInputTest.php:
```diff
@@ -26,17 +26,17 @@ public function testConstructor()
         $input = new ArgvInput(['cli.php', 'foo']);
         $input->bind(new InputDefinition([new InputArgument('name')]));
-        $this->assertSame(['name' => 'foo'], $input->getArguments(), '->parse() parses required arguments');
+        $this->assertEquals(['name' => 'foo'], $input->getArguments(), '->parse() parses required arguments');
```
Why those changes ? Aren't those reverted #58455 ?

theofidry replied:
Indeed, my bad

stof commented on src/Symfony/Component/Console/Tests/Input/ArgvInputTest.php:
```diff
@@ -594,4 +594,260 @@ public static function provideGetRawTokensTrueTests(): iterable
+        ?InputDefinition $inputDefinition,
+        ArgvInput $input,
+        ?array $parsedOptions,
+        array $expected,
```
Please keep the signature on one line. We use multiline signatures only in case we have promoted properties.

chalasr commented:
Ready for another approval @symfony/mergers

OskarStark commented on src/Symfony/Component/Console/Input/Input.php:
```diff
@@ -171,4 +191,61 @@ public function getStream()
+     *
+     * InputArguments must NOT be escaped as otherwise passing them to a `Process` would result in them being escaped twice.
+     *
+     * @param string[] $optionNames Name of the options returned. If empty, all options are returned and non-passed or non-existent are ignored.
```
```suggestion
     * @param string[] $optionNames Names of the options returned. If empty, all options are returned and non-passed or non-existent are ignored.
```

OskarStark commented on src/Symfony/Component/Console/Tests/Input/InputTest.php:
```diff
@@ -22,29 +22,30 @@ class InputTest extends TestCase
     {
         $input = new ArrayInput(['name' => 'foo'], new InputDefinition([new InputArgument('name')]));
-        $this->assertSame('foo', $input->getArgument('name'), '->__construct() takes a InputDefinition as an argument');
+        $this->assertEquals('foo', $input->getArgument('name'), '->__construct() takes a InputDefinition as an argument');
```
spult be reverted I guess

theofidry commented:
Addressed the last pieces of feedback and rebased it.

I also updated the description. A TL:DR; for a usage in the wild is https://github.com/webmozarts/console-parallelization/pull/328 which shows the hacky boilertemplate used to achieve this.

GromNaN commented on src/Symfony/Component/Console/Input/Input.php:
```diff
@@ -171,4 +191,62 @@ public function getStream()
+        $unparsedOptions = [];
+
+        foreach ($filteredRawOptions as $optionName => $parsedOption) {
+            $unparsedOption = self::unparseOption(
+                $this->definition->getOption($optionName),
+                $optionName,
+                $parsedOption,
+            );
+
+            $unparsedOptions[] = \is_array($unparsedOption) ? $unparsedOption : [$unparsedOption];
```
I think the code is simpler if you remove the private functions and you directly create arrays of strings:
```suggestion
            $option = $this->definition->getOption($optionName);
            $unparsedOptions[] = match (true) {
                $option->isNegatable() => [\sprintf('--%s%s', $value ? '' : 'no-', $name)],
                !$option->acceptValue() => [\sprintf('--%s', $name)],
                $option->isArray() => array_map(fn ($item) => \sprintf('--%s=%s', $name, $item), $value),
                default => [\sprintf('--%s=%s', $name, $value)],
            };
```

theofidry replied:
I personally find it harder to understand (the removing the private function part) but as you wish. Totally agree with making it an array right away though thanks

GromNaN commented on src/Symfony/Component/Console/Input/Input.php:
```diff
@@ -171,4 +191,62 @@ public function getStream()
         return $this->stream;
     }
+
+    /**
+     * Returns a stringified representation of the options passed to the command.
+     * The options must NOT be escaped as otherwise passing them to a `Process` would result in them being escaped twice.
+     *
+     * @param string[] $optionNames Names of the options returned. If null, all options are returned. Requested options
+     *                              that either do not exist or were not passed (even if the option has a default value)
+     *                              will not be part of the method output.
+     *
+     * @return list<string>
+     */
+    public function unparse(?array $optionNames = null): array
```
This function should be extracted to a dedicated class (SRP) so that it's easier to maintain: we can change it more easily. I don't know for the class name and namespace (we need to think about it). The method signature would be:

```php
function format(RawInputInterface $input, ?array $optionNames = null): array { /* ... */ }
```

Note: `webmozarts/console-parallelization` has a class dedicated to serialization, and that's great.

theofidry replied:
I agree on principle, and more generally `Input` and its children are doing too much making them harder to maintain (although thankfully they don't change often, so that alleviates that).

I am not sure it is best to move this now though, because:

- in essence it is the counter-part of `::parse()` which is part of `Input`, hence you would put two very closely related pieces in different places
- `::parse()` is abstract, hence every children implement its own. So in theory you could have different implementations of `::unparse()` to specific to an input implementation.

GromNaN commented on src/Symfony/Component/Console/Input/InputInterface.php:
```diff
@@ -18,6 +18,10 @@
  *
  * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @method getRawArguments(bool $strip = false): array<string|bool|int|float|null|array<string|bool|int|float|null>> Returns all the given arguments NOT merged with the default values.
+ * @method getRawOptions(): array<string|bool|int|float|array<string|bool|int|float|null>|null> Returns all the given options NOT merged with the default values.
+ * @method unparse(): list<string> Returns a stringified representation of the options passed to the command.
  */
 interface InputInterface
```
To avoid any BC break, we can add a new interface for these new capabilities:

```php
interface RawInputInterface
{
    public function getRawArguments(): array:
    public function getRawOptions(): array;
}
```

theofidry replied:
Would that really help? In the end what you receive, because that is what is typed everywhere, is `InputInterface`. Having a separate interface would bring back something similar to this code:

```php
    public static function getRawOptions(InputInterface $input): array
    {
        return $input instanceof RawInputInterface
            ? $input->getRawOptions()
            : [];
    }
```

Which... err yes in practice if we make `Input` implement it it will work 99% of the cases but type-wise leaves it to be desired.

theofidry replied:
Just to clarify: the only affected users are users implementing `InputInterface` without extending `Input`, which AFAIK (but I can be wrong) is _extremely_ rare, unless you're aware of more cases/usages in the wild?

GromNaN commented:
Sorry to be late in the review. I think it's a good feature, but the implementation can be improved.

GromNaN commented on src/Symfony/Component/Console/Input/Input.php:
```diff
@@ -171,4 +191,62 @@ public function getStream()
+     *
+     * @return list<string>
+     */
+    public function unparse(?array $optionNames = null): array
```
According to how you would use this method [in `webmozarts/console-parallelization`](https://github.com/webmozarts/console-parallelization/pull/328/files#diff-420c522b1de9b02d1f3c4163aaaabea35b6c12332f6c888738dff3cfc1226001R75-R76), you have a list of option names to exclude. This parameter should be `$excludedOptions`?

theofidry replied:
In `WebmozartsConsoleParallelization` we know the options we do NOT want, hence to get the ones we need to forward we need to juggle with the command definition. It is a specific use case for that feature, but not the only one, for instance you could want those values for specific options.

That's why I thought it would be better in Symfony to have a more re-usable piece of code (which is also simpler) even thought that means a tiny bit more work on my side.

nicolas-grekas reviewed (approved):
I rebased on 8.1, fixed the CS to follow the conventions used in this repo (no `use function` typically). I also split the new methods in `RawInputInterface` which extends `InputInterface` to skip a needless deprecation.
